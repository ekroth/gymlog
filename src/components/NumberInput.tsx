import React from 'react'
import { Button, Grid, Col, Input, Icon } from 'native-base'

type NumberInputProps = {
  onChange: (value: number) => void
  value: number
  step: number
  minValue: number
  type: 'number' | 'real'
}

export const NumberInput = (props: NumberInputProps) => (
  <Grid style={{ alignItems: 'center' }}>
    <Col style={{ minWidth: 20 }}>
      <Button
        style={{ alignSelf: 'center', backgroundColor: 'grey' }}
        onPress={() =>
          props.onChange(Math.max(props.minValue, props.value - props.step))
        }
      >
        <Icon name={'remove'} />
      </Button>
    </Col>
    <Col size={4}>
      <Input
        style={{
          alignSelf: 'stretch',
          fontSize: 40,
          textAlign: 'center',
          textAlignVertical: 'center'
        }}
        returnKeyType={'done'}
        defaultValue={`${props.value}`}
        keyboardType={'decimal-pad'}
        onSubmitEditing={event => {
          const n =
            props.type === 'number'
              ? parseFloat(event.nativeEvent.text)
              : parseInt(event.nativeEvent.text, 10)

          if (!Number.isNaN(n)) {
            props.onChange(n)
          } else {
            props.onChange(props.value)
          }
        }}
      />
    </Col>
    <Col style={{ minWidth: 20 }}>
      <Button
        style={{ alignSelf: 'center', backgroundColor: 'green' }}
        onPress={() => props.onChange(props.value + props.step)}
      >
        <Icon name={'add'} />
      </Button>
    </Col>
  </Grid>
)
