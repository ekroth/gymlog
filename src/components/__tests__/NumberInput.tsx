import React from 'react'
import {
  NativeSyntheticEvent,
  TextInputSubmitEditingEventData
} from 'react-native'
import { shallow } from 'enzyme'
import { NumberInput, NumberInputProps } from '../NumberInput'
import { Input } from 'native-base'

describe('SetHandlerComponent', () => {
  describe('rendering', () => {
    it('should match snapshot', () => {
      const props: NumberInputProps = {
        minValue: 0,
        step: 1,
        type: 'real',
        value: 100
      }
      const wrapper = shallow(<NumberInput {...props} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should call callback when receiving valid values otherwise old value', () => {
      const callback = jest.fn(() => {
        return
      })

      const event1 = {
        nativeEvent: {
          text: '120'
        }
      } as NativeSyntheticEvent<TextInputSubmitEditingEventData>

      const event2 = {
        nativeEvent: {
          text: 'abc'
        }
      } as NativeSyntheticEvent<TextInputSubmitEditingEventData>

      const props: NumberInputProps = {
        minValue: 0,
        onChange: callback,
        step: 1,
        type: 'real',
        value: 120
      }
      const wrapper = shallow(<NumberInput {...props} />)
      const input = wrapper.find(Input).first()

      input.props().onSubmitEditing!(event1)
      input.props().onSubmitEditing!(event2)

      expect(callback.mock.calls).toEqual([[120], [120]])
    })
  })
})
