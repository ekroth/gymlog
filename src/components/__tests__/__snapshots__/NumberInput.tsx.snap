// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetHandlerComponent rendering should match snapshot 1`] = `
<GridNB
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <ColumnNB
    style={
      Object {
        "minWidth": 20,
      }
    }
  >
    <Styled(Button)
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "grey",
        }
      }
    >
      <Styled(Icon)
        name="remove"
      />
    </Styled(Button)>
  </ColumnNB>
  <ColumnNB
    size={4}
  >
    <Styled(Input)
      defaultValue="100"
      keyboardType="decimal-pad"
      onSubmitEditing={[Function]}
      returnKeyType="done"
      style={
        Object {
          "alignSelf": "stretch",
          "fontSize": 40,
          "textAlign": "center",
          "textAlignVertical": "center",
        }
      }
    />
  </ColumnNB>
  <ColumnNB
    style={
      Object {
        "minWidth": 20,
      }
    }
  >
    <Styled(Button)
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "green",
        }
      }
    >
      <Styled(Icon)
        name="add"
      />
    </Styled(Button)>
  </ColumnNB>
</GridNB>
`;
